src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs

```diff
@@ -0,0 +1,156 @@
+        public static readonly string SourceIPAddressKey = "ProxyProtocolV2SourceIPAddress";
+        public static readonly string DestinationIPAddressKey = "ProxyProtocolV2DestinationIPAddress";
+        public static readonly string SourcePortKey = "ProxyProtocolV2SourcePort";
+        public static readonly string DestinationPortKey = "ProxyProtocolV2DestinationPort";
+        public static readonly string LinkIdKey = "ProxyProtocolV2LinkId";
+
+        /// <summary>
+        /// Proxy Protocol v2: https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt Section 2.2
+        /// Preamble(12 bytes) : 0D-0A-0D-0A-00-0D-0A-51-55-49-54-0A
+        ///  -21                        Version + stream    12
+        ///  -11                        TCP over IPv4       13
+        ///  -00-14                     length              14
+        ///  -AC-1C-00-04               src address         16
+        ///  -01-02-03-04               dest address        20
+        ///  -D7-9A                     src port            24
+        ///  -13-88                     dest port           26
+        ///  -EE                        PP2_TYPE_AZURE      28
+        ///  -00-05                     length              29
+        ///  -01                        LINKID type         31
+        ///  -33-00-00-26               LINKID              32.
+        /// </summary>
+        public static async Task ProcessAsync(ConnectionContext connectionContext, Func<Task> next, ILogger logger = null)
+        {
+            var input = connectionContext.Transport.Input;
+            while (true)
+            {
+                var result = await input.ReadAsync();
+                var buffer = result.Buffer;
+
+                try
+                {
+                    if (result.IsCompleted)
+                    {
+                        return;
+                    }
+
+                    // Buffer does not have enough data to make decision
+                    if (buffer.Length < 12)
+                    {
+                        continue;
+                    }
+
+                    var bufferArray = buffer.ToArray();
+                    if (!IsProxyProtocol(bufferArray))
+                    {
+                        // Break if it is not PPv2.
+                        break;
+                    }
+                    else
+                    {
+                        if (HasEnoughPpv2Data(bufferArray))
+                        {
+                            ExtractPpv2Data(ref buffer, bufferArray, connectionContext, logger);
+
+                            // It is PPv2, and we have enough data for PPv2
+                            break;
+                        }
+
+                        // It is PPv2, and we don't have enough data for PPv2
+                    }
+                }
+                finally
+                {
+                    input.AdvanceTo(buffer.Start);
```

Comment:
If this receives a partial proxy preamble, it could spin indefinitely.

```suggestion
                    input.AdvanceTo(buffer.Start, buffer.End);
```

https://docs.microsoft.com/en-us/dotnet/standard/io/pipelines#pipereader-common-problems

Reply:
This also means we'll probably have to have a loop inside the try/finally that calls ExtractPpv2Data to ensure we consume everything we can before saying we've examined the entire buffere.

Reply:
I now see why we didn't want to advance the examined cursor:

1. We don't want to examine past the "Ppv2Data" for proxy protocol connections.
2. We don't want to examine anything in case IsProxyProtocol later returns false in a later iteration. 

If we cannot either extract all the Ppv2Data or determine that it's not a PPv2 connection after the first read, I think we're going to have to create a new IDuplexPipe that replaces the original Reader with a new one that doesn't have everything pre-examined and set it on ConnectionContext.Transport.

Reply:
I think we can advance examined so long as it's a partial match, and if it's not a partial match we can break out.